Nginx ssl

ucyo opened the pull request:
Handling traffic via nginx and gunicorn

- [x] Route http traffic from 80 to app at 5000
- [x] Remove all port bindings and handle traffic via nginx
- [x] Setup https traffic at 443
- [x] Reroute http traffic to https
- [x] Hardening of SSL
- [x] Add documentation about cert generation
- [x] Check output directly to stdout / stderr
- [x] Test in staging environment
- [x] Add gunicorn support for app instead of running flask in development
- [x] Further harden SSL settings 
- [ ] Nginx settings based on environment

Test via https://www.ssllabs.com/ssltest/analyze.html?d=usedgoodreads.com

ucyo committed:
Add nginx to handle http traffic

ucyo committed:
Add nginx to handle http traffic

ucyo committed:
Traffic will be handled by nginx

Therefore the flask port needs to be exposed by the dockerfile

ucyo committed:
Traffic will be handled by nginx

Therefore the flask port needs to be exposed by the dockerfile

ucyo committed:
Add example certificates for staging environment

ucyo committed:
Add example certificates for staging environment

ucyo committed:
Fix travis CI/CD

ucyo committed:
Fix travis CI/CD

ucyo committed:
Add support for ssl

ucyo committed:
Add support for ssl

ucyo committed:
Service IPv6

ucyo committed:
Service IPv6

ucyo committed:
Handle http2

ucyo committed:
Handle http2

ucyo committed:
Redirect everything to ssl

ucyo committed:
Redirect everything to ssl

ucyo committed:
Make server_name an flexible

ucyo committed:
Make server_name an flexible

ucyo committed:
Harden SSL ciphers and prefer server settings

ucyo committed:
Harden SSL ciphers and prefer server settings

ucyo committed:
Rename file to server

nginx.conf is actually used for main settings and server for each
individual site

ucyo committed:
Rename file to server

nginx.conf is actually used for main settings and server for each
individual site

ucyo committed:
Add default nginx.conf

ucyo committed:
Add default nginx.conf

ucyo committed:
Further hardening of ssl

ucyo committed:
Further hardening of ssl

daniel-j-h commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "python.pythonPath": "/home/ucyo/Environments/juplab/bin/python"
+}
```
I think .vscode should be on the gitignore; and then not checking this into git :hugs:

daniel-j-h replied:
ALSO Y U NO NEOVIM :scream_cat:

ucyo replied:
Damn! Like a noob... :facepalm: Not that there is anything wrong with that... :)

ucyo replied:
> ALSO Y U NO NEOVIM scream_cat

Be happy its not emacs :smiling_imp:

daniel-j-h commented on docker-compose.override.yaml:
```diff
@@ -13,3 +11,7 @@ services:
+
+  nginx:
+    volumes:
+      - ./nginx/certs:/certs
```
Does this mean we use nginx and https for local development, too?

ucyo replied:
Yes. Otherwise the local development would differ quite a bit from the staging/production. It is not only running nginx and https but also the whole port settings would have to be changed by exposing flask again instead of only nginx

daniel-j-h replied:
Okay makes sense!

daniel-j-h commented on docker-compose.production.yaml:
```diff
@@ -4,8 +4,10 @@ services:
+
+  nginx:
+    volumes:
+      - /etc/letsencrypt/live/usedgoodreads.com/:/certs
```
We should have some docs wrt. how to set this up on the remote host in the readme.

ucyo replied:
Added to the README in f051b0811cfc7296f8f086aaa90289d19564b457

daniel-j-h commented on docker-compose.staging.yaml:
```diff
@@ -4,8 +4,10 @@ services:
+
+  nginx:
+    volumes:
+      - ./nginx/certs:/certs
```
Staging also gets deployed right now, but to port 8080 - can we use the same certificates as on production, or why are we using the one from local development here?

ucyo replied:
That is true. Using now same certs in staging with be12887658bf4dde05f3e45f86b5b0c9024c023c

daniel-j-h commented on docker-compose.yaml:
```diff
@@ -28,6 +28,22 @@ services:
+      - env/.nginx.env
+    ports:
+      - "80:80"
+      - "443:443"
```
Not sure these ports should be exposed in this `docker-compose.yaml` - for example for staging we need port 8080, no?

So in this base `docker-compose.yaml` we maybe don't want to expose any ports, and then do it in
- .override.yaml for local development
- .staging.yaml for the staging env
- .production.yaml for the production env

ucyo replied:
Added in ce6a6e303ba84582282ee3ac9d3799330c945031

daniel-j-h commented on docker-compose.yaml:
```diff
@@ -28,6 +28,22 @@ services:
+    ports:
+      - "80:80"
+      - "443:443"
+    command: /bin/bash -c "envsubst '$$NGINX_HOST' < /etc/nginx/conf.d/conf.template > /etc/nginx/conf.d/default.conf && nginx -g 'daemon off;'"
```
had no idea about `envsubst` - nice :+1:

daniel-j-h commented on docker-compose.yaml:
```diff
@@ -28,6 +28,22 @@ services:
+    depends_on:
+      - app
+    # #TODO Without this dependency the configuration of nginx can not use `app`
+    # as a hostname. Why?
```
Not sure here, but if we proxy traffic the app we need the app container to be up.

And potentially we should wait-for-it on app:5000 (?)

ucyo replied:
Yeah. I'm also unsure about this. But remove it, and suddenly nginx can not resolve the network `app` anymore

daniel-j-h replied:
So leave it in? :)

daniel-j-h commented on nginx/certs/fullchain.pem:
```diff
@@ -0,0 +1,24 @@
+g+FlQa/MMP5xpnCgzpiylPa3Tpj0AqAC5o7IMJgFo5xKVfZFlRe73wPWkzU/wUNw
+MAEDpbu0vmM682encg2M/cLYSClmppUYOu8wTQ5lOeGzwmJVbuM/78C4RZLZlncI
+0Yx3Us4+3lZcQJy4SIOnxMTBPBX8udyvxFWUkd8M8w==
+-----END CERTIFICATE-----
```
We should have some docs how we generated these :) Or a script for re-producibility.

ucyo replied:
Added to the todo list :rocket:

daniel-j-h commented on nginx/nginx.conf:
```diff
@@ -0,0 +1,37 @@
+                      '$status $body_bytes_sent "$http_referer" '
+                      '"$http_user_agent" "$http_x_forwarded_for"';
+
+    access_log  /var/log/nginx/access.log  main;
```
Are these not the defaults, or why do we need to specify e.g. logging, mime types, etc.?

ucyo replied:
Yes, all are defaults.  The custom settings start at line 33

daniel-j-h commented on nginx/nginx.conf:
```diff
@@ -0,0 +1,37 @@
+    access_log  /var/log/nginx/access.log  main;
+
+    sendfile        on;
+    #tcp_nopush     on;
```
I don't understand this

ucyo replied:
Me neither. It is default setting with nginx

daniel-j-h commented on nginx/server.conf:
```diff
@@ -0,0 +1,33 @@
+  ssl_ciphers EECDH+CHACHA20:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:EECDH+3DES:RSA+3DES:!MD5;
+
+  access_log /var/log/nginx/access.log;
+  error_log /var/log/nginx/error.log;
```
Quick question, we run a single nginx container. When we do `docker logs` on nginx, will we see anything? Because we log to these files here. Should we log to stdout / stderr in the nginx container, and then `docker logs` will show the nginx logs?

ucyo replied:
Yes, requests and fails are currently output via `docker log`. I can look if settings these to stderr will give us more information

ucyo replied:
Did some research. This is actually done by the [official nginx docker images](https://github.com/nginxinc/docker-nginx/blob/2ef3fa66f2a434cd5e44e35a02f4ac502cf50808/mainline/buster/Dockerfile#L96), no need for changes.

daniel-j-h reviewed (commented):
Nice! :clap: Left some smaller comments inline :arrow_up:

ucyo committed:
Remove vscode traces

ucyo committed:
Remove vscode traces

ucyo committed:
Use proper certificates at staging like in production

ucyo committed:
Use proper certificates at staging like in production

ucyo committed:
Add README information about self-signed certificates

ucyo committed:
Add README information about self-signed certificates

ucyo committed:
Fix typo

ucyo committed:
Fix typo

ucyo committed:
Declare ports for local development

ucyo committed:
Declare ports for local development

ucyo committed:
Declare proper ports for staging and production

ucyo committed:
Declare proper ports for staging and production

ucyo committed:
Install Gunicorn with gevent for async workers

ucyo committed:
Install Gunicorn with gevent for async workers

ucyo committed:
Remove `usedgoodreads` and use `gunicorn` for starting app

ucyo committed:
Remove `usedgoodreads` and use `gunicorn` for starting app

ucyo committed:
Gunicorn settings (better explicit than implicit)

- Log to stderr and stdout at level info
- Use two worker for asynch answering of request

ucyo committed:
Gunicorn settings (better explicit than implicit)

- Log to stderr and stdout at level info
- Use two worker for asynch answering of request

ucyo commented:
Using asynchronous `gunicorn` for starting `app` with bc03e61023f699bcf9436d5e17fd7ee614ca50a7

ucyo committed:
Add diffie-hellman settings for further hardening of SSL

ucyo committed:
Add diffie-hellman settings for further hardening of SSL

ucyo committed:
Only allow certain HTML Operations

ucyo committed:
Only allow certain HTML Operations

ucyo committed:
Prevent cross site attacks

ucyo committed:
Prevent cross site attacks

daniel-j-h commented on README.md:
````diff
@@ -42,8 +42,19 @@ This makes sure our development is self-contained and reproducible.
+## SSL settings
 
-## Infrastrcture
+The self-signed certificates currently in [nginx/certs](./nginx/certs) are generated using the following command:
+```
+openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 -keyout nginx/certs/fullchain.pem -out nginx/certs/privkey.pem
+```
+The Diffie-Hellman parameters are created using
+```
+openssl dhparam -out dhparams.pem 4096
+```
+Feel free to override these with the ones generated by yourself.
````
Thanks for this, this is gold :100: !!

daniel-j-h commented on docker-compose.override.yaml:
```diff
@@ -13,3 +11,10 @@ services:
+    volumes:
+      - ./nginx/certs:/certs
+    ports:
+      - "80:80"
+      - "443:443"
```
Should we really use these ports for local development? Why not e.g. 5000 like we had before?

ucyo replied:
1. Laziness. Why type `localhost:5000` if you can type `localhost` :)
2. Nginx can not handle the traffic if we set `5000:80` because the redirect would again go to `5000`. Basically `5000 > 80 > 443 > 5000 > ...`
3. Exposing `5000:5000` does also not work. `nginx` is not reachable than
(see comment below regarding 872c38b6395e8bd4ce1e08dcf5c8b1a8a254b92a)

daniel-j-h commented on docker-compose.production.yaml:
```diff
@@ -4,8 +4,13 @@ services:
+    volumes:
+      - /etc/letsencrypt/live/usedgoodreads.com/:/certs
+    ports:
+      - 80:80
+      - 443:443
```
Make it strings please

https://docs.docker.com/compose/compose-file/#ports

> When mapping ports in the HOST:CONTAINER format, you may experience erroneous results when using a container port lower than 60, because YAML parses numbers in the format xx:yy as a base-60 value. For this reason, we recommend always explicitly specifying your port mappings as strings.

ucyo replied:
Fixed in cedc53afb99cdda9d67c86db6ac8f7ebb2c7c887

daniel-j-h commented on docker-compose.staging.yaml:
```diff
@@ -4,8 +4,13 @@ services:
+    volumes:
+      - /etc/letsencrypt/live/usedgoodreads.com/:/certs
+    ports:
+      - 8080:80
+      - 4443:443
```
Same here

ucyo replied:
Fixed in cedc53afb99cdda9d67c86db6ac8f7ebb2c7c887

daniel-j-h commented on docker-compose.staging.yaml:
```diff
@@ -4,8 +4,13 @@ services:
+    volumes:
+      - /etc/letsencrypt/live/usedgoodreads.com/:/certs
+    ports:
+      - 8080:80
+      - 4443:443
```
Have you tried this actually? How does the browser know to request https from 4443?

ucyo replied:
Yes. I requested `localhost:4443` directly (see comment below regarding 872c38b6395e8bd4ce1e08dcf5c8b1a8a254b92a)

daniel-j-h commented on env/.nginx.env.example:
```diff
@@ -0,0 +1 @@
+NGINX_HOST=localhost
```
Why do we need this, and how does it have to look like for staging / production? Our domain? Or still localhost?

ucyo replied:
Currently this does not get changed. On staging and production this should be `usedgoodreads.com`, but that has other problems. See comment below regarding commit 872c38b6395e8bd4ce1e08dcf5c8b1a8a254b92a

daniel-j-h commented on nginx/server.conf:
```diff
@@ -0,0 +1,42 @@
+  ssl_dhparam         /certs/dhparams.pem;
+  ssl_prefer_server_ciphers on;
+  ssl_ciphers EECDH+CHACHA20:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:EECDH+3DES:RSA+3DES:!MD5;
+  ssl_protocols TLSv1.3;
```
This can make problems e.g. on ipads - we might want to use the intermediate config from here

https://ssl-config.mozilla.org/

ucyo replied:
wow. didn't know mozilla offered this. Fixed in e2057698a839588b2c60c0e9a267c7c8f9945c02

ucyo committed:
Apply recommendation from https://docs.docker.com/compose/compose-file/#ports

ucyo committed:
Apply recommendation from https://docs.docker.com/compose/compose-file/#ports

ucyo committed:
Updated using Mozilla recommendations

  - from https://ssl-config.mozilla.org/

ucyo committed:
Updated using Mozilla recommendations

  - from https://ssl-config.mozilla.org/

ucyo committed:
Change hostname and ports specifically for each environment

- This commit enables the use of different hostnames and ports for staging,
  production and local development with one nginx configuration.
- For using the staging instance, the ssl port (4443) must be accessed
  directly. The problem is that there is no way to knowledge to redirect
  shtml traffic to any other port than 443. Using two environments
  (staging and production) with different ssl ports with redirects on
  the same machine is not possible.

ucyo committed:
Change hostname and ports specifically for each environment

- This commit enables the use of different hostnames and ports for staging,
  production and local development with one nginx configuration.
- For using the staging instance, the ssl port (4443) must be accessed
  directly. The problem is that there is no way to knowledge to redirect
  shtml traffic to any other port than 443. Using two environments
  (staging and production) with different ssl ports with redirects on
  the same machine is not possible.

ucyo commented:
I currently can not see how we can run staging and production run on the same server with ssl. 
nginx must be able to redirect the traffic correctly to each instance. More information is in commit message: 872c38b6395e8bd4ce1e08dcf5c8b1a8a254b92a

Maybe we do not use SSL on staging?

daniel-j-h committed:
Cleans up unused kleinanzeigen prototype

ucyo committed:
Update gevent
